File: 125.验证回文串.py
#
# @lc app=leetcode.cn id=125 lang=python
#
# [125] 验证回文串
#
# 解法1(T40% S96%)：双指针判断各自指向的是否为相同的字符（如果是数字则相等，如果是字母则不考虑大小写，可以统一使用.lower()判断），因为除了数字和字母都要跳过，所以在while内部还要加两个while来过滤left和right值，可以使用.isalnum()判断是否为字母或数字

# @lc code=start
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        left, right = 0, len(s)-1
        while left < right:
            while left < right and not s[left].isalnum(): left += 1
            while left < right and not s[right].isalnum(): right -= 1
            if s[left].lower() != s[right].lower(): return False
            left += 1
            right -= 1
        return True
# @lc code=end

File: javascript/125.验证回文串.js
/*
 * @lc app=leetcode.cn id=125 lang=javascript
 *
 * [125] 验证回文串
 */

// @lc code=start
/**
 * @param {string} s
 * @return {boolean}
 */
var isPalindrome = function(s) {
    var isNum = function(ch) {
        if (ch>='0' && ch<='9') { return true; }
        else { return false; }
    }
    var isAlpha = function(ch) {
        if ((ch>='a' && ch<='z') || (ch>='A' && ch<='Z')) { return true; }
        else { return false; }
    }
    
    var isPalindrome = function(s) {
        var i=0, j=s.length-1;
        while (i<j) {
            while ((i<j) && !(isNum(s[i]) || isAlpha(s[i]))) { i++; }
            while ((i<j) && !(isNum(s[j]) || isAlpha(s[j]))) { j--; }
            if (i>=j) { break; }
    
            if (isNum(s[i]) && isNum(s[j]) && s[i]==s[j]) {
                i++;
                j--;
            } else if (isAlpha(s[i]) && isAlpha(s[j]) && s[i].toLowerCase()==s[j].toLowerCase()) {
                i++;
                j--;
            } else {
                return false;
            }
        }
        return true;
    };
};
// @lc code=end

